server/src/main/java/org/elasticsearch/tracing/Tracer.java

```diff
@@ -0,0 +1,176 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tracing;
+
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Releasable;
+
+import java.util.Map;
+
+/**
+ * Represents a distributed tracing system that keeps track of the start and end
+ * of various activities in the cluster. Traces are composed of "spans", each of
+ * which represents some piece of work with a duration. Spans are nested to
+ * create a parent-child hierarchy.
+ * <p>
+ * You can open a span using
+ * {@link #onTraceStarted(ThreadContext, String, String, Map)}, and stop it using
+ * {@link #onTraceStopped(String)}, at which point the tracing system will queue
+ * the data to be sent somewhere for processing and storage.
+ * <p>
+ * You can add additional data to a span using the {@code setAttribute(...)}
+ * methods, e.g.  {@link #setAttribute(String, String, String)}. This allows you
+ * to attach data that is not available when the span is opened.
+ * <p>
+ * <b>Implementation node:</b> tracers have similar performance requirements to loggers,
+ * in that they may be called in performance-sensitive locations. For example, network
+ * requests must be queued on separate threads, in order to avoid blocking any
+ * calling threads. Care should be taken to avoid allocating an excessive number of
+ * objects.
+ */
+public interface Tracer {
+
+    /**
+     * Called when a span starts.
+     * @param threadContext the current context. Required for tracing parent/child span activity.
+     * @param id a unique identifier for the activity, and will not be sent to the tracing system. Add the ID
+     *           to the attributes if it is important
+     * @param name the name of the span. Sent to the tracing system
+     * @param attributes arbitrary key/value data for the span. Sent to the tracing system
+     */
+    void onTraceStarted(ThreadContext threadContext, String id, String name, Map<String, Object> attributes);
+
+    /**
+     * Called when a span ends.
+     * @param id an identifier for the span
+     */
+    void onTraceStopped(String id);
+
+    /**
+     * Some tracing implementations support the concept of "events" within a span, marking a point in time during the span
+     * when something interesting happened. If the tracing implementation doesn't support events, then nothing will be recorded.
+     * This should only be called when a trace already been started on the {@code traceable}.
+     * @param id an identifier for the span
+     * @param eventName the event that happened. This should be something meaningful to people reviewing the data, for example
+     *                  "send response", "finished processing", "validated request", etc.
+     */
+    void onTraceEvent(String id, String eventName);
```

Comment:
Related to the comment above on naming, this could just be "traceEvent" and the next one "traceException"?